Add MapDatasetMetaData container

AtreyeeS opened the pull request:
Addresses  #4792. 

This PR adds a meta container on the MapDataset object. 

1. The stacking is not done in place currently. 
2. The `optional` dict keys are just stacked - and then users can decide what to do with it 
3. Stacking multiple instruments raises a warning - change to error and forbid?
4. The GTI table is already present on the map dataset, not replicating it here
5. Maybe another property `maker` can be added? To know how this dataset was created, eg: the outcome of some maker process or created by hand by the user. Or is that provenance information @bkhelifi 
6. In https://github.com/gammapy/gammapy/blob/main/gammapy/maps/utils.py#L77 we define INVALID_VALUE.int = np.nan
However, `isinstance(np.nan, int)` returns `False`, which raises `ValidationError`. I have put -999 as default for the event types, any better solution?

opinions please @registerrier @adonath @QRemy @maxnoe 

Still needs in future PRs:

- [ ] Deprecate the meta table?
- [ ] MapDatasetMaker.create_metadata()
- [ ] FluxPointsDataset meta object

bkhelifi commented:
Thanks a lot, @AtreyeeS .
About 2, it is not clear how the metadata reduction will be made. In order to avoid an exponential growing of metadata, I would not stack them per default now. I propose to think about them later when exchanging about metata reduction (with @kosack and @maxnoe ).
About 3, I would have say that this is forbidden, when instrument is not None. I do not see a reasonable use case for which it will be possible. However, one could change this behaviour in a future PR.
About 5, I think that this is indeed a provenance information.

For information, I am working on the `FluxMetaData` container. Once, we will have the main containers of the DL3 and DL4, one could work on the makers... the same for the estimators when we will have the main one for the DL4 and DL5.

bkhelifi commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+        The specific telescope subarray
+    observation_mode : str
+        observing mode
+    pointing : ~astropy.coordinates.SkyCoord
```
I think that this is not proper to a dataset, but to an individual observation. Is not it?

bkhelifi commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+        Telescope pointing direction
+    obs_ids : int
+        Observation ids stacked in the dataset
+    event_types : int
```
What is missing is the origin of the data, the data release name associated to the dir of the datastore.

AtreyeeS replied:
This will be copied from the `ObservationMetaData`?

bkhelifi commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+        Telescope pointing direction
+    obs_ids : int
+        Observation ids stacked in the dataset
+    event_types : int
```
There are other types of selection that can be made beyond the event types, the zen, the gti, the azimuth... I am wondering how and where one can 'store' this information in the workflow. Any idea?

bkhelifi commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+        Observation ids stacked in the dataset
+    event_types : int
+        Event types used in analysis
+    optional : dict
```
I share your suggestion to not store the individuals GTIs, but I am wondering whether a TSTART and a TSTOP can not be still useful. Any thought?

AtreyeeS replied:
I guess it depends a bit on how this meta information will be used

bkhelifi commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+__all__ = ["MapDatasetMetaData"]
+
+
+class MapDatasetMetaData(MetaData):
```
Should we not create a class for the Dataset object? and having daughter classes with their specificities? This would avoid code duplication...

AtreyeeS replied:
Or a more general `ObjectMetaData` as @registerrier suggested

registerrier commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+    event_type: Optional[Union[int, list[int]]]
+    optional: Optional[dict]
+
+    @validator("creation")
```
I am not sure you need a validation here.
Technically, in general, I think the `CreatorMetaData` can be None until the object is serialized (unless the information is read from disk).

AtreyeeS replied:
I thought `MapDataset.empty()` can add the `CreatorMetaData` and the rest of the information filled in by the `MapDatasetMaker`

registerrier commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+                f"Incorrect pointing. Expect CreatorMetaData got {type(v)} instead."
+            )
+
+    @validator("instrument")
```
Here no validation is needed either as `str` is a standard type.

AtreyeeS replied:
Yes, this I will remove. I wanted to be sure we forbid a list of strings here

registerrier commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+    telescope: Optional[Union[str, list[str]]]
+    observation_mode: Optional[Union[str, list]]
+    pointing: Optional[Union[SkyCoord, list[SkyCoord]]]
+    obs_ids: Optional[Union[int, list[int]]]
```
Are we sure `obs_id` can only be `int`?

AtreyeeS replied:
I had the same question about event types as well.

registerrier commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,157 @@
+                f"Incorrect pointing. Expect SkyCoord got {type(v)} instead."
+            )
+
+    @validator("obs_ids", "event_type")
+    def validate_obs_ids(cls, v):
+        if v is None:
+            return -999
```
Why would you use a value like -999. It could stay None. No?

registerrier reviewed (commented):
Thanks @AtreyeeS .

Maybe a very general comment first. We already see that informations are replicated between many metadata objects.  Also, it seems useless to validate the same information in many different places.

This means that we need a higher granularity in metadata structures. Our `ObjectMetaData` classes will then mostly contain lower level metadata objects + some informations specific to the object.

Then stacking would mostly be building a list of e.g. `ObsInfoMetaData` and `EventMetaData`. 

It would be good to also identify entries that are specific to the `Dataset`, if there are. 
Maybe the `name`, possibly the `target_name`

AtreyeeS commented:
I updated to use pydantic 2.0. 

I think the present stacking logic is too repetitive. Can it be simplified @adonath ?

adonath commented:
Thanks a lot @AtreyeeS! Probably the following code structure is simpler?
```python
from pydantic import BaseModel, RootModel
from typing import List, Optional


class MapDatasetMetaDataItem(BaseModel):
    name: str
    event_type: Optional[str] = None


class MapDatasetMetaData(RootModel):
    root: List[MapDatasetMetaDataItem]
    instrument: Optional[str] = None
    creation: ...

    def __iter__(self):
        return iter(self.root)

    def __getitem__(self, item):
        # could use the obs id here
        return self.root[item]
    
    def to_table(self):
        pass

    def from_table(self):
        pass
        
    def stack(self, item):
        # just append the list here
        pass


meta = StackedMapDatasetMetaData(
    [
        {"name": "hess"},
        {"name": "magic"},
        {"name": "veritas"},
        {"name": "fact"},
        {"name": "cta"},
    ]
)

meta[3].name
```
Having parallel lists is usually not a good idea, as it leads to these repetitive code statements. Introducing two separate classes might be the better solution here.

bkhelifi commented:
If we follow the ideaof @adonath , maybe one should keep the same logics of naming scheme between the class and the metadata class, ie:
MapDatasetMetaData, MapDatasetsMetaData (then leading to a change of the name of the python file)

In the list of instruments, one could add hawc, fermilat, astri....

Otherwise, I am wondering whether this class with this content is more a DatasetMetaData class... It seems very generic to all dataset child classes. It not it?

AtreyeeS commented:
I added a `from_meta_table` function to create a meta object from a meta table, and added computation of the meta in makers.

So the current behaviour is that there is a mechanism 

1. to compute the meta container from a meta table, and the meta table keeps information on all dataset operations (stacking, to_image, slicing etc). The meta container just contains the creation info.
2. Create the meta container during data reduction using the meta table

registerrier reviewed (commented):
Thanks @AtreyeeS . I have left a few comments.

AtreyeeS commented:
This now uses the `ObsInfoMetaData`.
Ready for final review from side

AtreyeeS commented on gammapy/datasets/metadata.py:
```diff
@@ -0,0 +1,103 @@
+        "creation": "CREATION",
+        "event_types": "EVT_TYPE",
+        "optional": "OPTIONAL",
+        "pointing": "POINTING",
```
without adding `pointing`, `to_header` fails if `altaz` is not present

registerrier replied:
See #5075 and #5076.

registerrier commented on gammapy/datasets/tests/test_metadata.py:
```diff
@@ -0,0 +1,131 @@
+
+def test_mapdataset_metadata_stack():
+    input1 = {
+        "obs_info": ObsInfoMetaData(**{"obs_id": 111}),
```
```suggestion
        "obs_info": ObsInfoMetaData(obs_id=111),
```
might be simpler

registerrier reviewed (approved):
Thanks @AtreyeeS . Looks good to me! No further comment.

AtreyeeS commented:
I checked that the meta information is not serialised with the dataset object.
I can add that in a following PR.

registerrier commented on gammapy/utils/metadata.py:
```diff
@@ -193,9 +193,10 @@ class ObsInfoMetaData(MetaData):
     _tag: ClassVar[Literal["obs_info"]] = "obs_info"
 
-    obs_id: int
+    obs_id: str
```
Why is it changed to `str`?

AtreyeeS replied:
We discussed previously that obs_ids don't need to be necessarily integers, and can be strings, and thus the MapDatasetMetaData was taking strings for obs_id

Having a `str` accepts both `int` and `str` on input

registerrier replied:
Currently the gadf requirement is that this is an int. That's why I kept this constrain when introducing the `ObsInfoMetaData`. 
Unless there is a strong constraint for the `MapDatasetMetaData` I'd rather keep the same solution and update later if necessary. No?

AtreyeeS replied:
I have no opinions on this.. I can then change the behaviour for the MapDataset as well...

AtreyeeS replied:
@bkhelifi The current implementation allows to read `int` as seen in the tests

registerrier replied:
Indeed, but pydantic is changing the type to `str`. 
For now, I think it is simpler to stick to `int`.

bkhelifi commented:
A test is failing:
FAILED ../../.tox/py311-test-alldeps_noray/lib/python3.11/site-packages/gammapy/data/tests/test_observations.py::test_observation_read - AssertionError: assert '20136' == 20136
 +  where '20136' = <gammapy.data.observations.Observation object at 0x7febafbb75d0>.obs_id

AtreyeeS commented:
1. reverted obs_id to int
2. a minimal serialisation of the meta implemented by extending the PRIMARY header
3. the serialisation does not work correctly when `obs_info` is a `list of ObsInfoMetaData` - I can try to fix that for 1.3

registerrier reviewed (approved):
Thanks @AtreyeeS . This looks good! No further comment from my side.

registerrier commented:
I have corrected an issue with `PointingInfoMetaData.from_header()` directly on your branch @AtreyeeS . The issue was that the default value was set to `SkyCoord(np.nan, no.nan)` which was breaking fits serialization. 
This is now solved with default values set to `None` if the header keywords are not defined.

registerrier reviewed (approved):
Thanks @AtreyeeS . Let's wait for the CI to run and merge.

bkhelifi commented:
This PR is associated to the issue #4791
